conans/model/conan_file.py

```diff
@@ -262,7 +263,9 @@ def _run():
                                              msys_mingw=msys_mingw, with_login=with_login)
         if run_environment:
             with tools.run_environment(self):
-                if OSInfo().is_macos:
+                if OSInfo().is_macos and isinstance(command, string_types):
```

Comment:
Similar to above, does this mean that sequences of commands will not prepend DYLD_LIBRARY_PATH? In that case shared libraries in OSX will not work with sequence of args commands?

Reply:
This is actually a direct consequence of going through the shell: the dynamic linker on macOS clears `DYLD_LIBRARY_PATH` when calling system programs in `/bin` or `/usr/bin`, so when using `shell=True` only the shell gets its setting, but not the called program.

However, when passing a sequence, the shell is not involved: therefore, the called command *does* get the value of `DYLD_LIBRARY_PATH` and the dynamic linker will use it to find dynamic libraries.

I did run a small workalike of `env` to verify that it indeed gets the `DYLD_LIBRARY_PATH` setting when invoked this way.

Reply:
IMO, we need a test for this one: just a recipe using `self.run` with a command that prints the environment would be enough, too many Python versions and the CI will help to check it works on all of them.

Reply:
I'm going to add a test to prove this point

Reply:
Added (and merged into this branch)